substrate/frame/revive/src/wasm/runtime.rs

```diff
@@ -1224,7 +1290,33 @@ pub mod env {
 		value_ptr: u32,
 		value_len: u32,
 	) -> Result<u32, TrapReason> {
-		self.set_storage(memory, flags, key_ptr, key_len, value_ptr, value_len)
+		self.set_storage(
+			memory,
+			flags,
+			key_ptr,
+			key_len,
+			StorageValue::Memory { ptr: value_ptr, len: value_len },
+		)
+	}
+
+	/// Sets the storage at a fixed 256-bit key with a fixed 256-bit value.
+	/// See [`pallet_revive_uapi::HostFn::set_storage_or_clear`].
+	#[stable]
+	#[mutating]
+	fn set_storage_or_clear(
+		&mut self,
+		memory: &mut M,
+		flags: u32,
+		key_ptr: u32,
+		value_ptr: u32,
+	) -> Result<u32, TrapReason> {
+		let value = memory.read_array::<32>(value_ptr)?;
```

Comment:
Saves a copy by directly reading the value onto the heap.

```suggestion
		let value = memory.read(value_ptr, 32)?;
```

Reply:
makes sense! thanks